sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,569 @@
+   */
+
+  def evaluateInSet(
+      attrRef: AttributeReference,
+      hSet: Set[Any],
+      update: Boolean)
+    : Option[Double] = {
+    if (!mutableColStats.contains(attrRef.exprId)) {
+      logDebug("[CBO] No statistics for " + attrRef)
+      return None
+    }
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+    val aType = attrRef.dataType
+
+    // use [min, max] to filter the original hSet
+    val validQuerySet = aType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, aType).asInstanceOf[NumericRange]
+        val hSetMap = hSet.map(e =>
+            if (e.isInstanceOf[String]) {
+              val utf8String = UTF8String.fromString(e.asInstanceOf[String])
+              numericLiteralToBigDecimal(aType, utf8String, StringType)
+            } else {
+              numericLiteralToBigDecimal(aType, e, aType)
+            })
+        hSetMap.filter(e => e >= statsRange.min && e <= statsRange.max)
+
+      // We assume the whole set since there is no min/max information for String/Binary type
+      case StringType | BinaryType => hSet
+    }
+    if (validQuerySet.isEmpty) {
+      return Some(0.0)
+    }
+
+    // newNdv should be no greater than the old ndv.  For example, column has only 2 values
+    // 1 and 6. The predicate column IN (1, 2, 3, 4, 5). validQuerySet.size is 5.
+    val newNdv = math.min(validQuerySet.size.toLong, ndv.longValue())
+    val(newMax, newMin) = aType match {
+      case _: NumericType =>
+        val tmpSet: Set[Double] = validQuerySet.map(e => e.toString.toDouble)
+        (Some(tmpSet.max), Some(tmpSet.min))
+      case DateType =>
+        if (hSet.forall(e => e.isInstanceOf[String])) {
+          val dateMax = Date.valueOf(hSet.asInstanceOf[Set[String]].max)
+          val dateMin = Date.valueOf(hSet.asInstanceOf[Set[String]].min)
+          (Some(dateMax), Some(dateMin))
+        } else {
+          val tmpSet: Set[Long] = validQuerySet.map(e => e.toString.toLong)
+          (Some(tmpSet.max), Some(tmpSet.min))
+        }
+      case TimestampType =>
+        if (hSet.forall(e => e.isInstanceOf[String])) {
+          val dateMax = Timestamp.valueOf(hSet.asInstanceOf[Set[String]].max)
+          val dateMin = Timestamp.valueOf(hSet.asInstanceOf[Set[String]].min)
+          (Some(dateMax), Some(dateMin))
+        } else {
+          val tmpSet: Set[Long] = validQuerySet.map(e => e.toString.toLong)
+          (Some(tmpSet.max), Some(tmpSet.min))
+        }
+      case _ =>
+        (None, None)
+    }
+
+    if (update) {
+      val newStats = attrRef.dataType match {
+        case _: NumericType | DateType | TimestampType =>
+          aColStat.copy(distinctCount = newNdv, min = newMin,
+            max = newMax, nullCount = 0)
+        case StringType | BinaryType =>
+          aColStat.copy(distinctCount = newNdv, nullCount = 0)

```

Comment:
actually we can remove this pattern match, for string/binary type, max/min will be None anyway